src/main/scala/org/constellation/util/MajorityStateChooser.scala

```diff
@@ -3,55 +3,100 @@ package org.constellation.util
 import io.chrisdavenport.log4cats.slf4j.Slf4jLogger
-import org.constellation.primitives.Schema._
+import org.constellation.domain.redownload.ReDownloadPlan
 import org.constellation.schema.Id
 import org.constellation.storage.RecentSnapshot
 
-import scala.util.{Random, Try}
-
-class MajorityStateChooser[F[_]: Concurrent] {
+import scala.util.Try
 
+object MajorityStateChooser {
   type NodeSnapshots = (Id, Seq[RecentSnapshot])
   type SnapshotNodes = (RecentSnapshot, Seq[Id])
 
-  val logger = Slf4jLogger.getLogger[F]
+  def compareSnapshotState(major: (Seq[RecentSnapshot], Set[Id]), ownSnapshots: List[RecentSnapshot]) = SnapshotDiff(
+    ownSnapshots.diff(major._1).sortBy(-_.height),
+    major._1.diff(ownSnapshots).toList.sortBy(-_.height),
+    major._2.toList
+  )
+
+  def chooseMajNodeIds(snapshotNodes: SnapshotNodes, nodeSnapshots: List[NodeSnapshots]): Option[Seq[Id]] =
+    nodeSnapshots.filter { case (id, recentSnapshotSeq) => snapshotNodes._2.contains(id) }.sortBy {
+      case (id, recentSnapshotSeq) => recentSnapshotSeq.size
+    }.map { case (id, recentSnapshotSeq) => id } match {
+      case Nil                   => None
+      case majNodIds @ (id :: _) => Some(majNodIds)
+    }
 
-  private final val differenceInSnapshotHeightToReDownloadFromLeader = 10
+  def planReDownload(allProposals: Map[Id, Map[Long, RecentSnapshot]],
+                     peers: Seq[Id],
+                     id: Id): ReDownloadPlan = {
+    val allProposalsNormalized = allProposals.toSeq.flatMap {
+      case (id, proposals: Map[Long, RecentSnapshot]) =>
+        proposals.toSeq.map { case (height, recentSnapshot) => (id, Seq(recentSnapshot)) }
+    }
+    val majority = MajorityStateChooser.chooseMajorWinner(peers, allProposalsNormalized)
+    val groupedProposals = allProposalsNormalized
+      .groupBy { case (id, recentSnapSeq) => id }
+      .map { case (id, tupList) => (id, tupList.flatMap(_._2)) }
+    val snapsThroughMaj = majority.map { case (height, (snap, nodes)) => MajorityStateChooser.getAllSnapsUntilMaj((snap, nodes), groupedProposals.toList) }
+    val majNodeIds =
+      majority.flatMap(maj => MajorityStateChooser.chooseMajNodeIds(maj._2, allProposalsNormalized.toList))
+    val (sortedSnaps, nodeIdsWithSnaps) =
+      (snapsThroughMaj.map(_.sortBy(-_.height)).getOrElse(Seq()), majNodeIds.map(_.toSet).getOrElse(Set()))
+    val diff = MajorityStateChooser.compareSnapshotState((sortedSnaps, nodeIdsWithSnaps), allProposals.getOrElse(id, Map()).values.toList)
+    ReDownloadPlan(id, sortedSnaps, nodeIdsWithSnaps, diff, groupedProposals)
+  }
+
+  def selectMajSnap(totalPeers: Int, allSnapshotNodes: Seq[SnapshotNodes]) = {
+    val numNodesWithSnapAtThisHeight = allSnapshotNodes.flatMap { case (recentSnap, ids) => ids }.length
+    val snapVotes = allSnapshotNodes.map { case (recSnap, ids) => (recSnap, ids.length) }
+    val simpleMaj = snapVotes.find { case (recSnap, count) => count - (totalPeers / 2) >= 0 }.map {
```

Comment:
Have you unit tested the case when `totalPeers` is odd number? Just trying to find some edge cases :p

Reply:
Nope, have not and we should. That will actually break, a count of 2 and totalPeers 5  will yield 0, which is incorrect since 2/5 is not a simple majority